include/AIToolbox/MDP/Model.hpp

```diff
@@ -259,7 +259,8 @@ namespace AIToolbox::MDP {
              *
              * @return A tuple containing a new state and a reward.
              */
-            std::tuple<size_t, double> sampleSR(size_t s, size_t a) const;
+            template <typename ST = size_t, typename AT = size_t>
```

Comment:
I don't really thing that MDP Model should be templatized. It has one task, and it is to model a single-agent MDP with constant actions, and it does this well. It's fine to design another class if you need a different thing. I'm not sure what these templates are trying to solve.

Reply:
It's not only for a non-stationary set of actions but also for actions (and/or states) that cannot be represented as just a ```size_t```, even if they always the same. Also, given that the types have defaults, not specifying them is exactly the current behaviour. Alternatively, we can have something like: this version of ```Model``` becomes ```BaseModel``` and the current ```Model``` (in your master) will just be a default implementation of the BaseModel.

Reply:
As methods that take models are templatized, having a `BaseModel` has no particular utility - the interfaces are defined with the `is_model` type trait. `CooperativeModel` in the factored part of the library can be used to represent state/actions as vectors, and in general algorithms that take that type of structure should go there, not in the `MDP` part of the library. The whole assumption in the `MDP` folder is that states and actions are `size_t`, and all algorithms and policies expect that. Doing a single exception for MCTS does not feel like the right choice.

Reply:
It took me time but I think I got it now. Because of the name and the usage, I thought ```model``` was supposed to be a base class. But it's in fact a default ```is_model``` complying implementation if you only care about stationary stuff, right? So I understand better why it shouldn't be templatized. 
The question that arrises though is why did you prefer have type traits instead of abstract classes given that it doesn't compile if it doesn't comply, much like a virtual pure method. I'll put the model back to where it was, it should resolve the comment just below as well.

Reply:
The thing is, using a base class with virtual methods doesn't really give you anything if you don't plan to use the class in a polymorphic way (which with a model you generally shouldn't). On the other hand, it slows down your code, as virtual calls are slower, and prevents the compiler from optimizing the code by inlining things. It makes the class hierarchy harder to understand, as it adds coupling between classes. It's also harder to integrate existing code, as for example you can't add an inheritance to a class you don't control (here it doesn't really apply but it's the general principle of it).

There is generally no reason to use inheritance and/or virtual inheritance if you are not getting anything back for it. Policies otoh use it, so that they can be easily composable/swappable. There it makes sense, so I did it :)

Reply:
But yeah, you are correct. Feel free to add any insight you have into the documentation!!